src/main/java/com/nimbusds/jose/JWSObjectJSON.java

```diff
@@ -0,0 +1,219 @@
+     * {@link State#SIGNED signed}.
+     *
+     * @param protectedHeader   The protected JWS header.
+     *                          Must not be {@code null}.
+     * @param unprotectedHeader The Per-Signature unprotected header.
+     *                          Might be {@code null}
+     * @param payload           The payload. Must not be {@code null}.
+     * @param signature         The signature. Must not be {@code null}.
+     *
+     * @throws ParseException If parsing of the serialised parts failed.
+     */
+    public JWSObjectJSON(Base64URL protectedHeader, UnprotectedHeader unprotectedHeader, Base64URL payload, Base64URL signature) throws ParseException {
+        super(protectedHeader, payload, signature);
+
+        this.unprotectedHeader = unprotectedHeader;
+    }
+
+    /**
+     * Returns Per-Signature Unprotected Header.
+     * Supports only ONE signature. Might be {@code null}
+     *
+     * @return Per-Signature Unprotected Header
+     */
+    public UnprotectedHeader getUnprotectedHeader() {
+        return unprotectedHeader;
+    }
+
+    @Override
+    public Map<String, Object> toJSONObject(boolean flattened) {
+        ensureSignedOrVerifiedState();
+
+        Map<String, Object> json = JSONObjectUtils.newJSONObject();
+
+        String signatureStr = getSignature().toString();
+        byte[] payload = getPayload().toBytes();
+        byte[] header = JSONObjectUtils.toJSONString(getHeader().toJSONObject(), true)
+                .getBytes(StandardCharsets.UTF_8);
+
+        if (flattened) {
+            json.put("protected", Base64URL.encode(header).toString());
+            json.put("payload", Base64URL.encode(payload).toString());
+            json.put("signature", signatureStr);
+        } else {
+            List<Map<String, Object>> signatures = new ArrayList<>();
+            Map<String, Object> signature = JSONObjectUtils.newJSONObject();
+            signature.put("protected", Base64URL.encode(header).toString());
+            signature.put("signature", signatureStr);
+
+            if (unprotectedHeader != null) {
+                signature.put("header", unprotectedHeader.toJSONObject());
+            }
+
+            signatures.add(signature);
+
+            json.put("payload", Base64URL.encode(payload).toString());
+            json.put("signatures", signatures);
+        }
+
+        return json;
+    }
+
+    /**
+     * Serialises this JWS object to JSON format.
+     *
+     * @return The serialised JWS object.
+     *
+     * @throws IllegalStateException If the JWS object is not in a
+     *                               {@link JWEObjectJSON.State#ENCRYPTED encrypted} or
+     *                               {@link JWEObjectJSON.State#DECRYPTED decrypted
+     *                               state}.
+     */
+    @Override
+    public String serialize() {
+
+        ensureSignedOrVerifiedState();
+
+        return JSONObjectUtils.toJSONString(toJSONObject(false), true);
```

Comment:
Personally I don't like boolean flag parameters because it's hard to tell from a call site what's going on - with the similarity in these methods I assumed they were both regarding `safe4web` - but I see it's not the case for `toJSONObject`! Also setting the flatten flat to `true` currently blows up with `NotImplementedException`.

Personally, I'd have separated it into two methods - `toJSONObjectFlattened` and `toJSONObject`, but this convention is used elsewhere in the library, so it's up to you!  If flattening isn't yet implemented and won't be before raising a PR upstream, perhaps it's worth removing entirely.  If you intend to implement it before raising a PR upstream, then please ignore me :D